packages/MonorepoBuilder/README.md

```diff
@@ -190,6 +190,17 @@ parameters:
     enable_default_release_workers: false
 ```
 
+### 8. Execute Commands Foreach contained Package
+
+While working with multiple packages in a monorepo you may feel the need to run arbitrary commands for each contained package.
+For example, you may want to run the tests for each package in isolation:
+
+```bash
+vendor/bin/monorepo-builder foreach -- sh -c 'composer install && vendor/bin/phpunit'
```

Comment:
This syntax look really confusing to me.

Better would be clear command or running it as own worker

Reply:
I can't follow you, do you mean release worker?
I don't see how using them would help with this. Could you explain a bit more how using workers would help here?

Maybe, explaining my problem which I try to solve with this command, will help:

My problem which is solved by the foreach command, is dealing with incomplete composer dependencies for the contained packages. So, in order to check if all dependencies are correct and each package works on its own, I have to to perform the following steps for each contained package:

```
# add path repositories in order to use dev versions of other contained packages
composer config repositories.local-packages '{"type": "path", "url": "../*", "options": {"symlink":true}}'
# install dependencies for this single package
composer install
# run tests for this single package
vendor/bin/phpunit
# cleanup
rm -rf composer.lock vendor
# remove path repositories
composer config repositories.local-packages --unset
```

I think this is the only way to ensure that each contained package is fully functional on its own. Lets say you have two packages `package-a` and `package-b`, both depending on `symfony/console` but only `package-a` has `symfony/console` defined in its requirements. If I just test with `vendor/bin/phpunit packages/package-b/phpunit.xml` everything works just fine because `symfony/console` is pulled into monorepo dependencies because of `package-a`. But as soon as I release `package-b` and only want to use `package-b`, it will break because `symfony/console` won't be installed.

And now think about having 20, 30 or even more packages in a monorepo, it is really easy to forget requiring a very common dependency because it may already be installed by a number of other packages.

One could add CI for each package, so it would be checked after splitting the repository but I think this is too late since the package is already published, so you have a broken master for all affected packages.

With the foreach command, you can simply perform those tasks for each package, no matter how many packages you have, this can be used during CI before splitting the repository and/or during development even before pushing the changes at all.

At least, this is the reason why I would like to have this kind of functionality.

However, there is still one problem with this approach, you can't fully cleanup the `composer.json` of each contained package, e.g. `composer config repositories.local-packages --unset` would leave `"repositories": {}` inside of each `composer.json`. To solve this, I'm working on an additional command for linking interdependencies, effectively adding path repositories for each dependency which is part of the monorepo. This command would also have the capability to fully cleanup the repositories section. I have a working prototype but I have still some smaller issues to solve.

With this additional command, one could use:
```
vendor/bin/monorepo-builder link-interdependencies
vendor/bin/monorepo-builder foreach composer install
vendor/bin/monorepo-builder foreach vendor/bin/phpunit
vendor/bin/monorepo-builder foreach -- rm -rf composer.lock vendor
vendor/bin/monorepo-builder link-interdependencies --unlink
```
or
```
vendor/bin/monorepo-builder link-interdependencies
vendor/bin/monorepo-builder foreach -- sh -c "composer install && vendor/bin/phpunit && rm -rf composer.lock vendor"
vendor/bin/monorepo-builder link-interdependencies --unlink
```


> This syntax look really confusing to me.

What is your problem with it?
If it is the need to use `--`, it really depends on the command you want to execute for each package. You have to tell SymfonyConsole that it should stop parsing arguments, otherwise it will throw an error of undefined options or take an option which should be passed to the command as its own option and weird things may happen. I think it's quite common to use `--` when multiple shell tools are used in combination.

However, if you don't feel comfortable with adding this to the MonorepoBuilder, just tell me, no problem. :wink: 
In this case, I will create my own utility package providing these commands, I just thought it may be a good addition.

Reply:
So your goal is to run tests on each package before splitting?

What I ment is to implement `ReleaseWorkerInterface` before splitting the repository. Only if tests would pass, it would go on.


If you can come with better syntax than `-- sh -c`, I'd be happy to consider it

Reply:
Yes, tests and before that, and I think even more important, `composer install` to make sure all required dependencies are defined within the respective `composer.json`.

I don't see this solely as part of the release cycle, although it would make sense there, too, but before any splitting is performed. Even without a new release, the repository may be split in order to provide updates for the current version being in development.

Getting rid of `-- sh -c` is easy if you use a shell script or composer for executing the commands, each package could define a script, e.g.:
```
    "scripts": {
        "test-single-package": [
            "composer config repositories.local-packages '{\"type\": \"path\", \"url\": \"../*\", \"options\": {\"symlink\":true}}'",
            "composer install",
            "vendor/bin/phpunit",
            "rm -rf composer.lock vendor",
            "composer config repositories.local-packages --unset"
        ]
    }
```

In this case `vendor/bin/monorepo-builder foreach composer test-single-package` would be sufficient. `--` is only required where options are used, e.g. for `sh -c` where otherwise the value of `-c` would be taken as the option for the MonorepoBuilder configuration file, so `--` has to be used to tell SymfonyConsole not to parse `-c` after `--` which was the reason for this pull request: https://github.com/Symplify/Symplify/pull/1608